doc/merkle.md

```diff
@@ -0,0 +1,139 @@
+# Merkle Structures
+
+MimbleWimble is designed for users to verify the state of the system given
+only pruned data. To achieve this goal, all transaction data is committed
+to the blockchain by means of Merkle trees which should support efficient
+updates and serialization even when pruned.
+
+Also, almost all transaction data (inputs, outputs, excesses and excess
+proofs) have the ability to be summed in some way, so it makes sense to
+treat Merkle sum trees as the default option, and address the sums here.
+
+A design goal of Grin is that all structures be as easy to implement and
+as simple as possible. MimbleWimble introduces a lot of new cryptography
+so it should made as easy to understand as possible. Its validation rules
+are simple to specify (no scripts) and Grin is written in a language with
+very explicit semantics, so simplicity is also good to achieve well-understood
+consensus rules.
+
+## Merkle Trees
+
+There are three Merkle trees committed to by each block:
+
+### Total Output Set
+
+Each object is one of two things: a (commitment, rangeproof) pair indicating
+an unspent output or a NULL marker indicating a spent one. It is a sum-tree
+over all unspent outputs (spent ones contribute nothing to the sum).
```

Comment:
Suppose I try to spend an output from block 1000 in block 1001. I make a transaction saying "spend output `abcd`, index 25000" and publish this, but then block 1000 gets replaced with one that makes output `abcd` have index 24999. Now my transaction is invalid, even though I didn't double-spend and nobody was acting maliciously. That's what I'm concerned about.

While its true that complete blocks don't have enough information for this sort of reorg to happen, in a model where miners do aggregation "at the last minute" and have some cache of unmerged transactions, this can certainly happen.